pkg/virt-controller/watch/drain/disruptionbudget/disruptionbudget.go

```diff
@@ -314,43 +315,90 @@ func (c *DisruptionBudgetController) execute(key string) error {
+	notFoundErrorCheck := func(err error) (bool, error) {
+		if errors.IsNotFound(err) {
+			return true, nil
+		}
+		return false, err
+	}
+
+	vmim, err := c.clientset.VirtualMachineInstanceMigration(vmi.Namespace).Get(migrationName, &v1.GetOptions{})
+	if err != nil {
+		return notFoundErrorCheck(err)
+	}
 
+	if vmim.IsFinal() {
+		listOptions := v1.ListOptions{
+			LabelSelector: fmt.Sprintf("%v=%v", virtv1.CreatedByLabel, vmi.UID),
+			FieldSelector: fmt.Sprintf("status.phase==%v", corev1.PodRunning),
+		}
+		pods, err := c.clientset.CoreV1().Pods(vmi.Namespace).List(context.Background(), listOptions)
+		if err != nil {
+			return notFoundErrorCheck(err)
+		}
+		return len(pods.Items) < 2, nil
+	}
+	return false, nil
 }
 
 func (c *DisruptionBudgetController) sync(key string, vmi *virtv1.VirtualMachineInstance, pdb *v1beta1.PodDisruptionBudget) error {
 	delete := false
 	create := false
+
 	// delete if there is no VMI
 	if vmi == nil && pdb != nil {
 		delete = true
 	} else if vmi != nil {
 		wantsToMigrate := wantsToMigrateOnDrain(vmi)
-		if vmi.DeletionTimestamp != nil && pdb != nil {
-			// pdb can already be deleted, shutdown already in process
-			delete = true
-		} else if !wantsToMigrate && pdb != nil {
-			// We don't want migrations on evictions, if there is a pdb, remove it
-			delete = true
-		} else if wantsToMigrate && vmi.DeletionTimestamp == nil && pdb == nil {
-			// No pdb and we want migrations on evictions
-			create = true
-		} else if wantsToMigrate && pdb != nil {
-			if ownerRef := v1.GetControllerOf(pdb); ownerRef != nil && ownerRef.UID != vmi.UID {
-				// The pdb is from an old vmi with a different uid, delete and later create the correct one
-				// The VMI always has a minimum grace period, so normally this should not happen, therefore no optimizations
+
+		if pdb != nil {
+			pdbMigrationName := pdb.ObjectMeta.Labels[virtv1.MigrationNameLabel]
+
+			if vmi.DeletionTimestamp != nil {
+				// pdb can already be deleted, shutdown already in process
+				delete = true
+			} else if pdbMigrationName != "" {
+				// only delete the temporary migration PDB
+				migrationCompleted, err := c.isMigrationComplete(vmi, pdbMigrationName)
+				if err != nil {
+					return err
+				}
+				delete = migrationCompleted
+			} else if !wantsToMigrate {
+				// We don't want migrations on evictions, if there is a pdb, remove it
 				delete = true
+			} else if wantsToMigrate {
+				if ownerRef := v1.GetControllerOf(pdb); ownerRef != nil && ownerRef.UID != vmi.UID {
+					// The pdb is from an old vmi with a different uid, delete and later create the correct one
+					// The VMI always has a minimum grace period, so normally this should not happen, therefore no optimizations
+					delete = true
+				}
 			}
+		} else if wantsToMigrate && vmi.DeletionTimestamp == nil {
+			// No pdb and we want migrations on evictions
+			create = true
 		}
 	}
-
 	if delete && pdb != nil && pdb.DeletionTimestamp == nil {
```

Comment:
maybe we can split the logic under this if and subsequent else if statements into two separate functions?

Reply:
done!